Create a WeakPtrFactory for use on the UI thread in VsyncWaiter

gw280 opened the pull request:
This fixes the failed assertion that the WeakPtr is being used on the wrong thread.

Issue: https://github.com/flutter/flutter/issues/44480

iskakaushik commented:
can we write a test for this?

gw280 commented:
It should be covered by the existing Shell unittests, we just don't run them yet on Fuchsia (currently on the roadmap)

gw280 commented:
Oh my bad, Shell unittests use their own implementation of VsyncWaiter for the test harness. I'll have a think about how best to do this but it may end up being put on the tail end of a long list of Fuchsia testing-related items as it'll likely need other work to be completed first (I'm thinking we can probably piggyback off the flutter_runner test harness work).

chinmaygarde commented on shell/platform/fuchsia/flutter/vsync_waiter.cc:
```diff
@@ -31,13 +34,30 @@ VsyncWaiter::VsyncWaiter(std::string debug_label,
   session_wait_.Cancel();
+
+  fml::AutoResetWaitableEvent ui_latch;
+  fml::MakeCopyable([weak_factory_ui = std::move(weak_factory_ui_),
```
Don't you need a |RunNowOrPostTask| here for the single threaded configuration?

gw280 replied:
Oh good catch. I copied the code from the equivalent code in Shell's ctor/dtor. Do we need to change Shell as well?

chinmaygarde replied:
I'd think so. Not sure how this wasn't covered by test already. Maybe add an explicit single threaded configuration to the shell_unittests (in a later patch is fine)?

gw280 replied:
Sorry, I'm not firing on all cylinders today. The Shell code does have it, I just accidentally omitted it.

iskakaushik commented:
@gw280 given that you have some subset of unittests already working on Fuchsia target device. How feasible is it to add a test to a new test suite that leverages this behavior.

gw280 commented:
> @gw280 given that you have some subset of unittests already working on Fuchsia target device. How feasible is it to add a test to a new test suite that leverages this behavior.

The unittests I'm currently looking at are only exercising the cross-platform code with stub implementations of things like VsyncWaiter so they unfortunately won't cover them.

Luckily though, @dnfield already did some work to add a11y tests at the flutter_runner layer (which this test would need to live in) and it does actually build the Fuchsia VsyncWaiter so it might not be too hard to add the test to that harness. I'll look into that now.

gw280 commented:
@iskakaushik unit test written and seems to work locally!

@chinmaygarde I wasn't thinking straight yesterday, I've updated the changes today and fixed a couple of issues (notably: i was constructing the UI thread's WeakPtrFactory inside a closure instead of the ctor itself, and I was dereffing the WeakPtrFactory on the platform thread in the lambda capture instead of on the UI thread). This has now been resolved and the test passes.

iskakaushik commented on shell/platform/fuchsia/flutter/vsync_waiter.cc:
```diff
@@ -57,7 +78,11 @@ void VsyncWaiter::AwaitVSync() {
-      [self = weak_factory_.GetWeakPtr()] {
+      [& weak_factory_ui = this->weak_factory_ui_] {
+        if (!weak_factory_ui) {
+          return;
```
we should maybe error log here.

iskakaushik commented on shell/platform/fuchsia/flutter/vsync_waiter.cc:
```diff
@@ -33,11 +36,29 @@ VsyncWaiter::VsyncWaiter(std::string debug_label,
+  // thread so we have ordering guarantees (see ::AwaitVSync())
+  fml::TaskRunner::RunNowOrPostTask(
+      task_runners_.GetUITaskRunner(), fml::MakeCopyable([this]() mutable {
+        this->weak_factory_ui_ =
```
@chinmaygarde should we block this call. I think its ok to not block but jut want to confirm.

chinmaygarde replied:
I think this is fine because of the reason mentioned in the comment.
